### PR TITLE
refine softmax fwd on CPU

### PR DESCRIPTION
**Pull request by tensor-tang**
improve 6X on v2620 v3

Before
image

After
image

### CODE REVIEW

**Comment by luotao1** [comment]
> If reason is that input is to small to vsEXP then  how about on moving vsEXP out of batch loop. vsEXP takes most of time  perhaps we could first in a loop (for all batches) compute max and sub and then vsEXP  for all data followed by loop executing sums and scals?

@jczaja Do you mean update this PR?

File: paddle/fluid/operators/math/cpu_vec.h [hunk 1]
```diff
@@ -128,6 +135,47 @@ inline void vec_scal<float, platform::avx512f>(const int n, const float a,
   vec_scal<float, platform::avx2>(n, a, x, y);
 }
 
+template <typename T, platform::cpu_isa_t isa = platform::isa_any>
+inline void vec_sum(const size_t n, const T* x, T* s) {
+  s[0] = x[0];
```
Comment (jczaja) [inline paddle/fluid/operators/math/cpu_vec.h:140]: Just small suggestion. It could be worth to add OpenMP SIMD vectorization (the way it is described in our article). Reasoning is that for sum reduction it gives a boost and works for SSE4 so code would benefit from it on older platform. Details in our article (https://arxiv.org/pdf/1904.12380.pdf)  Figure 6 (lines 27 - 31) in section 2.4.2 . That would work starting from GCC 4.9.2

File: paddle/fluid/operators/math/softmax_impl.h [hunk 2]
```diff
@@ -38,16 +40,37 @@ template <typename DeviceContext, typename T, bool is_test, typename Enable>
...
-
+  auto in_dims = X->dims();
   const int kBatchDim = 0;
   const int kClassDim = 1;
 
-  const int batch_size = logits.dimension(kBatchDim);
-  const int num_classes = logits.dimension(kClassDim);
+  const int num_classes = in_dims[kClassDim];
+  const int batch_size = in_dims[kBatchDim];
   const int num_remain = num_classes / axis_dim;
```
Comment (jczaja) [inline paddle/fluid/operators/math/softmax_impl.h]: axis_dim can be (to my understanding): [0, rank-1] . So it may happen that  there is exception here when dividing by zero . If my understanding is wrong then please explain?

File: paddle/fluid/operators/math/softmax_impl.h [hunk 3]
```diff
@@ -38,16 +40,37 @@ template <typename DeviceContext, typename T, bool is_test, typename Enable>
...
 
+  if (num_remain == 1 && platform::MayIUse(platform::avx)) {
```
Comment (jczaja) [inline paddle/fluid/operators/math/softmax_impl.h]: Could you please why num_remain has to be equal 1 ? I understand that efficient softmax code can be used when axis_dim == rank -1  eg. we noramlize after final dimension.  Why not has condition: axis_dim + 1 = in.dims.size()

File: paddle/fluid/operators/math/softmax_impl.h [hunk 4]
```diff
@@ -38,16 +40,37 @@ template <typename DeviceContext, typename T, bool is_test, typename Enable>
...
+
+      T sum = 0;
+      vec_sum<T, platform::avx>(num_classes, out_data, &sum);
+      sum = static_cast<T>(1) / sum;
+      vec_scal<T, platform::avx>(num_classes, sum, out_data, out_data);
+
+      in_data += num_classes;
+      out_data += num_classes;
+    }
+    return;
```
Comment (jczaja) [inline paddle/fluid/operators/math/softmax_impl.h]: What will happen when we have situation that we exectue softmax normalization after first axe or no AVX support. It seems there will no be any computation? Is this correct

**Review by jczaja — COMMENTED** [review]
@luotao1 That was just suggestion on how to implement softmax op to deal with small inputs (while batch are bigger). It is not directly related to this PR

File: paddle/fluid/operators/math/cpu_vec.h (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/cpu_vec.h:140]: goog point by `# pragma omp simd redution  (+ : sum)`

But will this use multi-threads? I am afraid we could only use single thread.

File: paddle/fluid/operators/math/softmax_impl.h (hunk 2 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/softmax_impl.h]: https://github.com/PaddlePaddle/Paddle/blob/dfdcd91869595c7f068767e15f05f325d8a1de9c/paddle/fluid/operators/softmax_op.h#L56

It's from here, so could no be zero.

File: paddle/fluid/operators/math/softmax_impl.h (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/softmax_impl.h]: It's refer to [here](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/jit/refer/refer.h#L444)

File: paddle/fluid/operators/math/softmax_impl.h (hunk 4 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/softmax_impl.h]: This only improve on avx platform and only axis=1, if not, use original implementations.

File: paddle/fluid/operators/math/cpu_vec.h (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/math/cpu_vec.h:140]: No. this is single threaded eg. omp simd does perform vectorization

File: paddle/fluid/operators/math/softmax_impl.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax_impl.h]: Ok, makes sense

File: paddle/fluid/operators/math/softmax_impl.h (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax_impl.h]: ok

**Review by jczaja — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/math/softmax_impl.h (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax_impl.h]: ok

File: paddle/fluid/operators/math/cpu_vec.h (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/cpu_vec.h:140]: OK thanks.

**Comment by jczaja** [comment]
@tensor-tang Just one more question: This code is to be run for inference only eg. it is enable when PaddlePaddle for inference is build and no ValueClip is there that is needed in softmax fwd training. But performance results are shown from training which should not contain this optimization ?

**Comment by tensor-tang** [comment]
This is used for training actually.

**Comment by tensor-tang** [comment]
http://ci.paddlepaddle.org/viewLog.html?buildId=101697&buildTypeId=Paddle_PrCi&tab=buildLog&branch_Paddle=pull%2F16684 CI failed on test_parallel_executor_seresnext

**Review by luotao1 — APPROVED** [review]
LGTM
